tests/k8st/tests/test_bgp_advert.py

```diff
@@ -143,6 +228,112 @@ def assert_ecmp_routes(self, dst, via=["10.192.0.3", "10.192.0.4"]):
           matchStr += "\n\tnexthop via %s  dev eth0 weight 1" % ip
         retry_until_success(lambda: self.assertIn(matchStr, self.get_routes()))
 
+    def test_rr(self):
+        self.tearDown()
+        self.setUpRR()
+
+        # Create ExternalTrafficPolicy Local service with one endpoint on node-1
+        run("""kubectl apply -f - << EOF
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-rr
+  namespace: bgp-test
+  labels:
+    app: nginx
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: nginx
+      run: nginx-rr
+  template:
+    metadata:
+      labels:
+        app: nginx
+        run: nginx-rr
+    spec:
+      containers:
+      - name: nginx-rr
+        image: nginx:1.7.9
+        ports:
+        - containerPort: 80
+      nodeSelector:
+        beta.kubernetes.io/os: linux
+        kubernetes.io/hostname: kube-node-1
+---
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx-rr
+  namespace: bgp-test
+  labels:
+    app: nginx
+    run: nginx-rr
+spec:
+  ports:
+  - port: 80
+    targetPort: 80
+  selector:
+    app: nginx
+    run: nginx-rr
+  type: NodePort
+  externalTrafficPolicy: Local
+EOF
+""")
+
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get nodes -o yaml")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get bgppeers -o yaml")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get bgpconfigs -o yaml")
+
+        # Update the node-2 to behave as a route-reflector
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get node kube-node-2 -o json > /home/node.json")
+        run("sed -i '11 a \ \ \ \ \ \ \"i-am-a-route-reflector\": \"true\",' /home/node.json")
+        run("sed -i '21 a \ \ \ \ \ \ \"routeReflectorClusterID\": \"224.0.0.1\",' /home/node.json")
```

Comment:
This works for now but is fragile in that it may break when new fields are added to the Node structure.  Instead I think we should:
- get the Node JSON as a string, from the `run` function:
  ```
  json_str = run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get node kube-node-2 -o json")
  ```
- convert that to a Python dict:
  ```
  node_dict = json.loads(json_str)
  ```
- set the fields that we need for RR operation:
  ```
  node_dict['metadata']['labels']['i-am-a-route-reflector'] = 'true'
  node_dict['spec']['bgp']['routeReflectorClusterID'] = '224.0.0.1'
  ```
- convert back to a string, and apply:
```
run("""kubectl exec -i -n kube-system calicoctl -- /calicoctl apply -f - << 'EOF'
%s
EOF
""" % json.dumps(node_dict))
```

Do you agree that that would be more robust?

Reply:
Thank you for suggesting this.